test/integration/hds_integration_test.cc

```diff
@@ -75,54 +67,438 @@ class HdsIntegrationTest : public HttpIntegrationTest,
+  host_stream->encodeData(1024, true);
+  hds_stream_->waitForGrpcMessage(*dispatcher_, response);
+  EXPECT_EQ(
+      envoy::api::v2::core::HealthStatus::HEALTHY,
+      response.mutable_endpoint_health_response()->mutable_endpoints_health(0)->health_status());
+  EXPECT_EQ(host_upstream_->localAddress()->ip()->port(),
+            response.mutable_endpoint_health_response()
+                ->mutable_endpoints_health(0)
+                ->mutable_endpoint()
+                ->mutable_address()
+                ->mutable_socket_address()
+                ->port_value());
+  EXPECT_EQ(host_upstream_->localAddress()->ip()->addressAsString(),
+            response.mutable_endpoint_health_response()
+                ->mutable_endpoints_health(0)
+                ->mutable_endpoint()
+                ->mutable_address()
+                ->mutable_socket_address()
+                ->address());
+
+  test_server_->waitForCounterGe("cluster.anna.health_check.success", 1);
+
+  // Clean up connections
+  host_fake_connection_->close();
+  host_fake_connection_->waitForDisconnect();
+  cleanupHdsConnection();
 
   EXPECT_EQ(1, test_server_->counter("hds_delegate.requests")->value());
   EXPECT_EQ(2, test_server_->counter("hds_delegate.responses")->value());
+  EXPECT_EQ(1, test_server_->counter("cluster.anna.health_check.success")->value());
+  EXPECT_EQ(0, test_server_->counter("cluster.anna.health_check.failure")->value());
+}
+
+TEST_P(HdsIntegrationTest, SingleEndpointTimeout) {
+  initialize();
+
+  // Messages
+  envoy::service::discovery::v2::HealthCheckRequest envoy_msg;
+  envoy::service::discovery::v2::HealthCheckRequestOrEndpointHealthResponse response;
+  envoy::service::discovery::v2::HealthCheckSpecifier server_health_check_specifier =
+      makeHealthCheckSpecifier();
+
+  // Server <--> Envoy
+  hds_fake_connection_ = hds_upstream_->waitForHttpConnection(*dispatcher_);
+  hds_stream_ = hds_fake_connection_->waitForNewStream(*dispatcher_);
+  hds_stream_->waitForGrpcMessage(*dispatcher_, envoy_msg);
+
+  // Server asks for healthchecking
+  hds_stream_->startGrpcStream();
+  hds_stream_->sendGrpcMessage(server_health_check_specifier);
+
+  EXPECT_EQ(host_upstream_->httpType(), FakeHttpConnection::Type::HTTP1);
+  host_fake_connection_ = host_upstream_->waitForHttpConnection(*dispatcher_);
+  test_server_->waitForCounterGe("hds_delegate.requests", ++hds_requests_);
+
+  // Envoy sends a healthcheck message to an endpoint
+  host_stream = host_fake_connection_->waitForNewStream(*dispatcher_);
+  host_stream->waitForEndStream(*dispatcher_);
+  EXPECT_STREQ(host_stream->headers().Path()->value().c_str(), "/healthcheck");
+  EXPECT_STREQ(host_stream->headers().Method()->value().c_str(), "GET");
+
+  // Endpoint doesn't repond to the healthcheck
+  hds_stream_->waitForGrpcMessage(*dispatcher_, response);
+  EXPECT_EQ(
+      envoy::api::v2::core::HealthStatus::UNHEALTHY,
```

Comment:
Yeah, I think the TIMEOUT status would be ideal here, can you add a TODO to handle this?